---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
options(stringsAsFactors = F)
```

Funtions for HumanMethylation450 y MethylationEPIC beadchips.


```{r}
library(missMethyl)
library(limma)
library(minfi)
library(minfiData)
```

Getting files to load.

```{r}
baseDir <- system.file("extdata", package = "minfiData")
targets <- read.metharray.sheet(baseDir)
```

Loading data. The function read.metharray.exp reads IDAT files. Two-color IDAT files ara pair of files with names ending in _Red.idat or Grn.idat.

```{r}
help("read.metharray.exp")
rgSet <- read.metharray.exp(targets = targets)
rgSet
```

Using the SWAN method substantially reduces the technical variability between the assay designs whilst maintaining important biological differences. The SWAN method makes the assumption that the number of CpGs within the 50bp probe sequence reflects the underlying biology of the region being interrogated. Hence, the overall distribution of intensities of probes with the same number of CpGs in the probe body should be the same regardless of assay type. The method then uses a subset quantile normalization approach to adjust the intensities of each array (Maksimovic, Gordon, and Oshlack 2012).

SWAN can take a MethylSet, RGChannelSet or MethyLumiSet as input. It should be noted that, in order to create the normalization subset, SWAN randomly selects Infinium I and II probes that have one, two and three underlying CpGs; as such, we recommend using set.seed before to ensure that the normalized intensities will be identical, if the normalization is repeated.

```{r}
set.seed(123)
```

proprocessRaw. Preprocesses raw data. This function takes the Read and the Green channel of an Illumina methylation array, together with its associated manifest object and converts it into a MethylSet containing the methylated and unmethtylated signal.

```{r}
help("preprocessRaw")
mSet <- preprocessRaw(rgSet)
```

The SWAN method has two parts. First, an average quantile distribution is created using a subset of probes defined to be biologically similar based on the number of CpGs underlying the probe body. This is achieved by randomly selecting N Infinium I and II probes that have 1, 2 and 3 underlying CpGs, where N is the minimum number of probes in the 6 sets of Infinium I and II probes with 1, 2 or 3 probe body CpGs. If no probes have previously been filtered out e.g. sex chromosome probes, etc. N=11,303. This results in a pool of 3N Infinium I and 3N Infinium II probes. The subset for each probe type is then sorted by increasing intensity. The value of each of the 3N pairs of observations is subsequently assigned to be the mean intensity of the two probe types for that row or 'quantile'. This is the standard quantile procedure. The intensities of the remaining probes are then separately adjusted for each probe type using linear interpolation between the subset probes.

```{r}
help("SWAN")
mSetSw <- SWAN(mSet,verbose=TRUE)
```

Plotting beta value distributions.Beta values (??) are the estimate of methylation level using the ratio of intensities between methylated and unmethylated alleles. ?? are between 0 and 1 with 0 being unmethylated and 1 fully methylated.

High-throughput profiling of DNA methylation status of CpG islands is crucial to understand the epigenetic regulation of genes. The microarray-based Infinium methylation assay by Illumina is one platform for low-cost high-throughput methylation profiling. Both Beta-value and M-value statistics have been used as metrics to measure methylation levels.

We demonstrate that the relationship between the Beta-value and M-value methods is a Logit transformation, and show that the Beta-value method has severe heteroscedasticity for highly methylated or unmethylated CpG sites. In order to evaluate the performance of the Beta-value and M-value methods for identifying differentially methylated CpG sites, we designed a methylation titration experiment. The evaluation results show that the M-value method provides much better performance in terms of Detection Rate (DR) and True Positive Rate (TPR) for both highly methylated and unmethylated CpG sites. 


```{r}
par(mfrow=c(1,2), cex=1.25)
densityByProbeType(mSet[,1], main = "Raw")
densityByProbeType(mSetSw[,1], main = "SWAN")
```

Methylation of cytosine bases in DNA CpG islands is an important epigenetic regulation mechanism in the organ development, aging and different disease statuses. Hypermethylation of CpG islands located in the promoter regions of tumor suppressor genes has been firmly established as one of the most common mechanisms for gene regulation in cancer [2, 3]. Therefore, high-throughput profiling of DNA methylation status of CpG islands is crucial for forwarding our understanding of the influence of epigenomics. Microarray-based Illumina Infinium methylation assay has been recently used in epigenomic studies due to its high throughput, good accuracy, small sample requirement and relatively low cost.To estimate the methylation status, the Illumina Infinium assay utilizes a pair of probes (a methylated probe and an unmethylated probe) to measure the intensities of the methylated and unmethylated alleles at the interrogated CpG site. The methylation level is then estimated based on the measured intensities of this pair of probes. To date, two methods have been proposed to measure the methylation level. The first one is called Beta-value, ranging from 0 to 1, which has been widely used to measure the percentage of methylation. This is the method currently recommended by Illumina. The second method is the log2 ratio of the intensities of methylated probe versus unmethylated probe [13]. We have referred to it as the M-value method because it has been widely used in the mRNA expression microarray analysis. Since both methods have their own strengths and limitations, understanding the performance characteristics of both measures is very important in providing the best methylation analysis.
The Beta-value is the ratio of the methylated probe intensity and the overall intensity (sum of methylated and unmethylated probe intensities). Following the notation used by Illumina methylation assay [12], Beta-value for an ith interrogated CpG site is defined as:
Betai=max(yi,methy,0)max(yi,unmethy,0)+max(yi,methy,0)+??	(1)
where y i,menty and y i,unmenty are the intensities measured by the ith methylated and unmethylated probes, respectively. To avoid negative values after background adjustment, any negative values will be reset to 0. Illumina recommends adding a constant offset ?? (by default, ?? = 100) to the denominator to regularize Beta value when both methylated and unmethylated probe intensities are low. The Beta-value statistic results in a number between 0 and 1, or 0 and 100%. Under ideal conditions, a value of zero indicates that all copies of the CpG site in the sample were completely unmethylated (no methylated molecules were measured) and a value of one indicates that every copy of the site was methylated. If we assume the probe intensities are Gamma distributed, then the Beta-value follows a Beta distribution. For this reason, it has been named the Beta-value.

The M-value is calculated as the log2 ratio of the intensities of methylated probe versus unmethylated probe as shown in Equation 2:
Mi=log2(max(yi,methy,0)+??max(yi,unmethy,0)+??)	(2)
Here we slightly modified the definition given in [13] by adding an offset ?? (by default, ?? = 1) to the intensity values to prevent unexpected big changes due to small intensity estimation errors, since for very small intensity values (especially between 0 and 1), small changes of the methylated and unmethylated probe intensities can result in large changes in the M-value. A M-value close to 0 indicates a similar intensity between the methylated and unmethylated probes, which means the CpG site is about half-methylated, assuming that the intensity data has been properly normalized by Illumina GenomeStudio or some other external normalization algorithm. Positive M-values mean that more molecules are methylated than unmethylated, while negative M-values mean the opposite. 

In high-throughput statistical data analyses, many of them, like canonical linear models or ANOVA, assume the data is homoscedastic, i.e., the variable variances are approximately constant. The violation of this assumption, which is described as heteroscedasticity in statistics, imposes serious challenges when applying these analyses to high-throughput data [15]. A common way to check the homoscedasticity of the data is by visualizing the relations between mean and standard deviation [15, 16]. Figure 3 shows the mean and standard deviation relations of the Beta-value and M-value, which were calculated based on technical replicates. The red dots represent the median standard deviation within a local window. The data was first ranked by mean methylation levels, and then binned into twenty non-overlapping windows, with each bin containing 5% of the data. The standard deviation of Beta-value is greatly compressed in the low (between 0 and 0.2) and high (between 0.8 and 1) ranges. This means Beta-value has significant heteroscedasticity in the low and high methylation range. The problem of heteroscedasticity is effectively resolved after transforming Beta-value to M-value using Equation 3. We can see M-value is approximately homoscedastic. 


Filtering poor quality probes

Poor quality probes can be filtered out based on the detection p-value. For this example, to retain a CpG for further analysis, we require that the detection p-value is less than 0.01 in all samples.

```{r}
mSetSw
detP <- detectionP(rgSet)
keep <- rowSums(detP < 0.01) == ncol(rgSet)
mSetSw <- mSetSw[keep,]
```

Extracting B and M values

Now that the data has been SWAN normalised we can extract ??
?? and M-values from the object. We prefer to add an offset to the methylated and unmethylated intensities when calculating M-values, hence we extract the methylated and unmethylated channels separately and perform our own calculation. For all subsequent analysis we use a random selection of 20000 CpGs to reduce computation time.

```{r}
mset_reduced <- mSetSw[sample(1:nrow(mSetSw), 20000),]
meth <- getMeth(mset_reduced)
unmeth <- getUnmeth(mset_reduced)
Mval <- log2((meth + 100)/(unmeth + 100))
beta <- getBeta(mset_reduced)
dim(Mval)
```

An MDS plot (Figure 2) is a good sanity check to make sure samples cluster together according to the main factor of interest, in this case, cancer and normal.

```{r}
par(mfrow=c(1,1))
plotMDS(Mval, labels=targets$Sample_Name, col=as.integer(factor(targets$status)))
legend("topleft",legend=c("Cancer","Normal"),pch=16,cex=1.2,col=1:2)
```

Testing for differential methylation using limma.

To test for differential methylation we use the limma package (G. K. Smyth 2005), which employs an empirical Bayes framework based on Guassian model theory. First we need to set up the design matrix. There are a number of ways to do this, the most straightforward is directly from the targets file. There are a number of variables, with the status column indicating cancer/normal samples. From the person column of the targets file, we see that the cancer/normal samples are matched, with 3 individuals each contributing both a cancer and normal sample. Since the limma model framework can handle any experimental design which can be summarised by a design matrix, we can take into account the paired nature of the data in the analysis. For more complicated experimental designs, please refer to the limma User's Guide.

```{r}
group <- factor(targets$status,levels=c("normal","cancer"))
id <- factor(targets$person)
design <- model.matrix(~id + group)
design
```

Now we can test for differential methylation using the lmFit and eBayes functions from limma. As input data we use the matrix of M-values.

```{r}
fit.reduced <- lmFit(Mval,design)
fit.reduced <- eBayes(fit.reduced)
```

The numbers of hyper-methylated (1) and hypo-methylated (-1) can be displayed using the decideTests function in limma and the top 10 differentially methylated CpGs for cancer versus normal extracted using topTable.

```{r}
summary(decideTests(fit.reduced))
top<-topTable(fit.reduced,coef=4)
top
```

Note that since we performed our analysis on M-values, the logFC and AveExpr columns are computed on the M-value scale. For interpretability and visualisation we can look at the ?? ?? values. The beta values for the top 4 differentially methylated CpGs shown in Figure 3.

```{r}
cpgs <- rownames(top)
par(mfrow=c(2,2))
for(i in 1:4){
stripchart(beta[rownames(beta)==cpgs[i],]~design[,4],method="jitter",
group.names=c("Normal","Cancer"),pch=16,cex=1.5,col=c(4,2),ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgs[i],cex.main=1.5)
}
```

Testing for differential variability.

Rather than testing for differences in mean methylation, we may be interested in testing for differences between group variances. For example, it has been hypothesised that highly variable CpGs in cancer are important for tumour progression (K. D. Hansen et al. 2011). Hence we may be interested in CpG sites that are consistently methylated in the normal samples, but variably methylated in the cancer samples.
In general we recommend at least 10 samples in each group for accurate variance estimation, however for the purpose of this vignette we perform the analysis on 3 vs 3. In this example, we are interested in testing for differential variability in the cancer versus normal group. Note that when we specify the coef parameter, which corresponds to the columns of the design matrix to be used for testing differential variability, we need to specify both the intercept and the fourth column. The ID variable is a nuisance parameter and not used when obtaining the absolute deviations, however it can be included in the linear modelling step. For methylation data, the function will take either a matrix of M-values, ??
?? values or a object as input. If ?? ?? values are supplied, a logit transformation is performed. Note that as a default, varFit uses the robust setting in the limma framework, which requires the use of the statmod package.

```{r}
fitvar <- varFit(Mval, design = design, coef = c(1,4))
summary(decideTests(fitvar))
topDV <- topVar(fitvar, coef=4)
topDV
```

```{r}
cpgsDV <- rownames(topDV)
par(mfrow=c(2,2))
for(i in 1:4){
stripchart(beta[rownames(beta)==cpgsDV[i],]~design[,4],method="jitter",
group.names=c("Normal","Cancer"),pch=16,cex=1.5,col=c(4,2),ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```

DVCs generally exhibited an outlier structure with the increased or decreased variability caused by large changes in DNA methylation present in only a small number of "outlier" samples. About 9.45% of DVCs (2,707 out of 28,653) were driven by single outliers (defined as only one sample displaying > 20% change and all the other samples displaying < 5% change in DNA methylation).3 As expected, DMVCs showed features of both DMCs and DVCs with relatively homogeneous differential methylation changes but larger variance in the obese group.
